src/src/PluginStructs/P145_data_struct.h

```diff
@@ -0,0 +1,106 @@
+// ----------------------------------------------------------------------------
+// P145 "Gases - MQxxx (MQ135 CO2, MQ3 Alcohol)"
+// Definition of sensor abstraction
+// See _P145_MQxxx.ino
+// 2023 By flashmark
+// ----------------------------------------------------------------------------
+#ifndef PLUGINSTRUCTS_P145_DATA_STRUCT_H
+#define PLUGINSTRUCTS_P145_DATA_STRUCT_H
+
+
+#include "../../_Plugin_Helper.h"
+#ifdef USES_P145
+
+// Maximum number of types that can be selected from
+#define P145_MAXTYPES  10
+
+// Analog input pin for sensor value
+#define P145_SENSOR_PIN A0
+
+// Supply voltage applied to the sensor
+#define P145_VCC        (5.0)
+// Max input scale voltage of the ADC
+#define P145_VMAX       (3.3)
+
+// Calibration interval in milli seconds (once per day)
+#define P145_CALIBRATION_INTERVAL (24*60*60*1000)
+
+// Selection of the conversion algorithm Rs/R0 => value
+enum P145_algorithm 
+{
+  p145AlgNone,  // Undefined/free algorithm
+  p145AlgA,     // MQ-135 with optional temp/hum correction 
+  p145AlgB,     // miguel5612 Exponential
+  p145AlgC      // miguel5612 Linear
+};
+
+// Structure with sensor specific conversion data
+// Members are preferrably sorted by memory alignment
+struct P145_SENSORDEF
+{
+  float cleanRatio;   // Rs/R0 ratio in clean air
+  float para;         // PARA scaling factor value
+  float parb;         // PARB exponent value
+  float cora;         // CORA
+  float corb;         // CORB
+  float corc;         // CORC
+  float cord;         // CORD
+  float core;         // CORE
+  float corf;         // CORF
+  float corg;         // CORG
+  P145_algorithm alg; // Preferred/tuned algorithm
+  char name[8];       // Sensor type name
+  char gas[8];        // Measured gas concentration
```

Comment:
No member is initialized during construction.
The char arrays can be initialized like this:
`char gas[8] = {0};`
The floats can be initialized like this:
`float para = 0.0f;`
or the more generic and new (since C++11) approach:
`float para{};` or `float para = {};`   which will also initialize it to the most sensible default value for these basic types which do not have their own constructor. (0.0f for floats)
N.B. for types which do have a constructor, calling `{}` or `={}` will actually call their constructor.
Usually for just members of some struct or class type, their constructor is being called. However for arrays of such a type it is not guaranteed their constructor is called.  Not sure if this is also behavior as is defined in the standard, but I've experienced this recently where members of some array were not initialized.

So better to always initialize explicitly. Either by assigning some value, or adding the `{}` when declaring them in a .h file.

Reply:
I fully agree with proper initialization. In this case the structure is explicitly defined to hold a table of predefined sensor types in PROGMEM. There is one copy (you may call it a cache) in P145_data_struct. This will pretty well secure integrity. However let's keep to the project's coding style.

Reply:
I need some help here as I am not a C++ expert. When I initialize teh fiels teh compiler complains about the hard coded "const struct P145_SENSORDEF sensorDefs[] PROGMEM" definition in P145_data_struct.cpp. I need a table with many sensors with each sensors characteristics and that data should be kept in PROGMEM. Only for the actual sensor the data is copied to RAM.

Reply:
OK, please help me remind to check this tomorrow morning...